### PR TITLE
ENH: Improve and export Slicer2AUT transform

### LINKED ISSUE #102: tra seem to be wrt dicom origin rather than 'autoscoper' origin
When generating VRG from a static volume or from a seqence, the Origin2DICOM transform represents relationship between dicom space and aut space.

| Dicom space |  Origin2DICOM used to express volume in AUT space 
|--|--|
| image | image |

:warning: AUT space (center of volume and VRGs) is the space from which .tra transforms should be expressed. Taking  .tra : the positions are not in appropriate anatomical location. See below.

| AUT space (center of volume and VRGs) is the space from which .tra transforms should be expressed. Taking  .tra : the positions are not in appropriate anatomical location | @NicerNewerCar pointed- they appear translationaly to be in DICOM basis: |
|--|--|
| image | image |

### LINKED ISSUE #107: Auto generated tra offset from correct projection
image
Using BN00107.

bone coloured volume rendering is a clone of neutral, with invert Origin2DICOM applied.

Yellow segmented stls

Grey - neutral volume rendering (dciom space)

Green in mc1 with (invert) tra applied

image

### PR DESCRIPTION
**Pull request by NicerNewerCar**
Closes #107 
Closes #102

### CODE REVIEW

File: AutoscoperM/AutoscoperMLib/IO.py [hunk 1]
```diff
@@ -255,56 +256,13 @@ def writeTFMFile(filename: str, spacing: list[float], origin: list[float]):
...
-        for j in range(4):  # Col
-            rowwise.append(str(transform.GetElement(i, j)))
-
-    with open(filename, "w+") as f:
-        f.write(",".join(rowwise))
+def writeTRA(fileName: str, transforms: list[vtk.vtkMatrix4x4]) -> None:
+    rowWiseStrings = []
+    for transform in transforms:
+        rowWiseStrings.append([str(transform.GetElement(i, j)) for i, j in product(range(4), range(4))])
+    with open(fileName, "w") as traFile:
```
Comment (NicerNewerCar) [inline AutoscoperM/AutoscoperMLib/IO.py:263]:
```suggestion
    with open(fileName, "w+") as traFile:
```

File: AutoscoperM/AutoscoperM.py [hunk 2]
```diff
@@ -1427,3 +1428,79 @@ def IsVolumeCentered(node: Union[slicer.vtkMRMLVolumeNode, slicer.vtkMRMLSequenc
...
+
+        slicer2Autoscoper = vtk.vtkMatrix4x4()
+        slicer2AutoscoperNode.GetMatrixTransformToParent(slicer2Autoscoper)
+
+        # Extract the rotation matrices so we are not affecting the translation vector
+        transformR = vtk.vtkMatrix3x3()
+        slicer2autoscoperR = vtk.vtkMatrix3x3()
+        for i, j in product(range(3), range(3)):
+            transformR.SetElement(i, j, transform.GetElement(i, j))
+            slicer2autoscoperR.SetElement(i, j, slicer2Autoscoper.GetElement(i, j))
```
Comment (jcfr) [inline AutoscoperM/AutoscoperM.py:1473]:
```suggesiton
        vtkAddon.vtkAddonMathUtilities.GetOrientationMatrix(transform, transformR)
        vtkAddon.vtkAddonMathUtilities.GetOrientationMatrix(slicer2Autoscoper, slicer2autoscoperR)
```

File: AutoscoperM/AutoscoperM.py [hunk 3]
```diff
@@ -1427,3 +1428,79 @@ def IsVolumeCentered(node: Union[slicer.vtkMRMLVolumeNode, slicer.vtkMRMLSequenc
...
+
+        vtk.vtkMatrix3x3.Multiply3x3(slicer2autoscoperR, transformR, transformR)
+
+        for i, j in product(range(3), range(3)):
+            transform.SetElement(i, j, transformR.GetElement(i, j))
```
Comment (jcfr) [inline AutoscoperM/AutoscoperM.py:1478]:
```suggestion
        vtkAddon.vtkAddonMathUtilities.SetOrientationMatrix(transformR, transform)

File: AutoscoperM/AutoscoperM.py [hunk 4]
```diff
@@ -1427,3 +1428,79 @@ def IsVolumeCentered(node: Union[slicer.vtkMRMLVolumeNode, slicer.vtkMRMLSequenc
...
+
+        # Apply the translation vector
+        for i in range(3):
+            transform.SetElement(i, 3, transform.GetElement(i, 3) + slicer2Autoscoper.GetElement(i, 3))
+
+        slicer.mrmlScene.RemoveNode(slicer2AutoscoperNode)
```
Comment (jcfr) [inline AutoscoperM/AutoscoperM.py:1484]: This should be moved in the function where the transform is effectively added.

File: AutoscoperM/AutoscoperM.py [hunk 5]
```diff
@@ -1427,3 +1428,79 @@ def IsVolumeCentered(node: Union[slicer.vtkMRMLVolumeNode, slicer.vtkMRMLSequenc
...
+
+        return transform
+
+    @staticmethod
+    def createSlicer2AutoscoperTransformNode(volumeNode: slicer.vtkMRMLVolumeNode) -> slicer.vtkMRMLLinearTransformNode:
```
Comment (jcfr) [inline AutoscoperM/AutoscoperM.py:1489]: -> `createAndAddSlicer2AutoscoperTransformNode`

File: AutoscoperM/AutoscoperM.py [hunk 6]
```diff
@@ -1427,3 +1431,76 @@ def IsVolumeCentered(node: Union[slicer.vtkMRMLVolumeNode, slicer.vtkMRMLSequenc
...
+
+    @staticmethod
+    def loadTransformFromFile(transformFileName: str) -> slicer.vtkMRMLLinearTransformNode:
+        return slicer.util.loadNodeFromFile(transformFileName)
+
+    @staticmethod
+    def applyOrigin2DicomTransform(
+        transform: vtk.vtkMatrix4x4,
+        origin2DicomTransformNode: slicer.vtkMRMLLinearTransformNode,
+        ApplyDicom2Origin: bool = False,
```
Comment (jcfr) [inline AutoscoperM/AutoscoperM.py:1443]:
```suggestion
        applyDicom2Origin: bool = False,
```

File: AutoscoperM/AutoscoperM.py [hunk 7]
```diff
@@ -1427,3 +1431,76 @@ def IsVolumeCentered(node: Union[slicer.vtkMRMLVolumeNode, slicer.vtkMRMLSequenc
...
+    ) -> vtk.vtkMatrix4x4:
+        if ApplyDicom2Origin:
```
Comment (jcfr) [inline AutoscoperM/AutoscoperM.py:1445]:
```suggestion
        if applyDicom2Origin:
```

File: AutoscoperM/AutoscoperM.py [hunk 8]
```diff
@@ -1427,3 +1431,76 @@ def IsVolumeCentered(node: Union[slicer.vtkMRMLVolumeNode, slicer.vtkMRMLSequenc
...
+        vtk.vtkMatrix4x4.Multiply4x4(origin2DicomTransformMatrix, transform, transform)
+
+        slicer.mrmlScene.RemoveNode(origin2DicomTransformNode)
+        return transform
+
+    @staticmethod
+    def applySlicer2AutoscoperTransform(
+        transform: vtk.vtkMatrix4x4,
+        slicer2AutoscoperNode: slicer.vtkMRMLLinearTransformNode,
+        ApplyAutoscoper2Slicer: bool = False,
```
Comment (jcfr) [inline AutoscoperM/AutoscoperM.py:1460]:
```suggestion
        applyAutoscoper2Slicer: bool = False,
```

File: AutoscoperM/AutoscoperM.py [hunk 9]
```diff
@@ -1427,3 +1431,76 @@ def IsVolumeCentered(node: Union[slicer.vtkMRMLVolumeNode, slicer.vtkMRMLSequenc
...
+    ) -> vtk.vtkMatrix4x4:
+        """Utility function for converting a transform between the Slicer and Autoscoper coordinate systems."""
+
+        if ApplyAutoscoper2Slicer:
```
Comment (jcfr) [inline AutoscoperM/AutoscoperM.py:1464]:
```suggestion
        if applyAutoscoper2Slicer:
```

File: AutoscoperM/AutoscoperM.py [hunk 10]
```diff
@@ -1427,3 +1431,76 @@ def IsVolumeCentered(node: Union[slicer.vtkMRMLVolumeNode, slicer.vtkMRMLSequenc
...
+    @staticmethod
+    def loadTransformFromFile(transformFileName: str) -> slicer.vtkMRMLLinearTransformNode:
+        return slicer.util.loadNodeFromFile(transformFileName)
+
+    @staticmethod
+    def applyOrigin2DicomTransform(
+        transform: vtk.vtkMatrix4x4,
+        origin2DicomTransformNode: slicer.vtkMRMLLinearTransformNode,
+        ApplyDicom2Origin: bool = False,
+    ) -> vtk.vtkMatrix4x4:
```
Comment (jcfr) [inline AutoscoperM/AutoscoperM.py:1444]:
```suggestion
    ) -> vtk.vtkMatrix4x4:
       """Utility function for ... ."""
```

File: AutoscoperM/AutoscoperM.py [hunk 11]
```diff
@@ -1427,3 +1431,76 @@ def IsVolumeCentered(node: Union[slicer.vtkMRMLVolumeNode, slicer.vtkMRMLSequenc
...
+        if ApplyDicom2Origin:
+            origin2DicomTransformNode.Inverse()
+
+        origin2DicomTransformMatrix = vtk.vtkMatrix4x4()
+        origin2DicomTransformNode.GetMatrixTransformToParent(origin2DicomTransformMatrix)
+
+        vtk.vtkMatrix4x4.Multiply4x4(origin2DicomTransformMatrix, transform, transform)
+
+        slicer.mrmlScene.RemoveNode(origin2DicomTransformNode)
```
Comment (jcfr) [inline AutoscoperM/AutoscoperM.py:1453]:
```suggestion
```

Should be moved in the function where the transform is originally added

File: AutoscoperM/AutoscoperM.py [hunk 12]
```diff
@@ -1427,3 +1431,76 @@ def IsVolumeCentered(node: Union[slicer.vtkMRMLVolumeNode, slicer.vtkMRMLSequenc
...
+            slicer2AutoscoperNode.Inverse()
+
+        slicer2Autoscoper = vtk.vtkMatrix4x4()
+        slicer2AutoscoperNode.GetMatrixTransformToParent(slicer2Autoscoper)
+
+        # Extract the rotation matrices so we are not affecting the translation vector
+        transformR = vtk.vtkMatrix3x3()
+        slicer2AutoscoperR = vtk.vtkMatrix3x3()
+        vtkAddon.vtkAddonMathUtilities.GetOrientationMatrix(transform, transformR)
+        vtkAddon.vtkAddonMathUtilities.GetOrientationMatrix(slicer2Autoscoper, slicer2AutoscoperR)
```
Comment (jcfr) [inline AutoscoperM/AutoscoperM.py:1474]:
```suggestion
        vtkAddon.vtkAddonMathUtilities.GetOrientationMatrix(transform, transformR)
        slicer2AutoscoperR = vtk.vtkMatrix3x3()
        vtkAddon.vtkAddonMathUtilities.GetOrientationMatrix(slicer2Autoscoper, slicer2AutoscoperR)
```

File: AutoscoperM/AutoscoperM.py [hunk 13]
```diff
@@ -1427,3 +1431,76 @@ def IsVolumeCentered(node: Union[slicer.vtkMRMLVolumeNode, slicer.vtkMRMLSequenc
...
+        for i in range(3):
+            transform.SetElement(i, 3, transform.GetElement(i, 3) + slicer2Autoscoper.GetElement(i, 3))
+
+        return transform
+
+    @staticmethod
+    def createAndAddSlicer2AutoscoperTransformNode(
+        volumeNode: slicer.vtkMRMLVolumeNode,
+    ) -> slicer.vtkMRMLLinearTransformNode:
+        # Slicer 2 Autoscoper Transform
```
Comment (jcfr) [inline AutoscoperM/AutoscoperM.py:1490]: Add docstring

**Review by jcfr — COMMENTED** [review]
After applying tweaks :+1:

**Pull request merged**